### PR TITLE
added standalone apex test runner target

### PR DESCRIPTION
**Pull request by zhili1208**

### CODE REVIEW

File: build/build.proj [hunk 1]
```diff
@@ -122,6 +122,33 @@
 
   <!--
     ============================================================
+    Run VS15 apex tests
+    ============================================================
+  -->
+  <Target Name="ApexTestsStandalone"  DependsOnTargets="RestoreApex">
```
Comment (emgarten) [inline build/build.proj]: Build usually can't run in the same msbuild process as restore, this might make it more confusing/not work.

Comment (nkolev92) [reply build/build.proj]: I think there's a way to do it now...I think runnig build with /restore does it? 
Basically it just cleans the context after restore. 
But as is currently it won't work i think

**Comment by nkolev92** [comment]
Where is the apex package currently? 
Is it public or on one of our feeds? 
How will it restore successfully?

File: build/build.proj (hunk 1 above)
Comment (zhili1208) [reply build/build.proj]: I tried this, didn't get any error. I can remove the dependsOnTarget, then we just need to restore before run

**Comment by zhili1208** [comment]
Apex package is still in private feed, it's not public, on CI, we need to override nuget.config to make it work

**Comment by rohit21agrawal** [comment]
Can you add a step in configure.ps1 such that if CI is set to true as an environment variable then it adds the source to the NuGet.config in our repository?
We can modify build definitions and store credentials as passwords which will be available as environment variables

**Review by rohit21agrawal — CHANGES_REQUESTED** [review]
Add step in configure.ps1 as per comment

File: configure.ps1 [hunk 2]
```diff
@@ -54,6 +54,12 @@ Invoke-BuildStep 'Installing .NET CLI' {
     Install-DotnetCLI -Force:$Force    
 } -ev +BuildErrors
 
+if ($Ci) {
```
Comment (rohit21agrawal) [inline configure.ps1]: change to if ($env:CI)

File: build/common.ps1 [hunk 3]
```diff
@@ -189,6 +189,13 @@ Function Install-NuGet {
     & $NuGetExe locals all -list -verbosity detailed
 }
 
+Function Add-ApexFeed {
+    if (Test-Path $NuGetExe) {    
+        # Add Apex VSTS feed
+        & $NuGetExe sources add -Name ApexFeed -Source "https://devdiv.pkgs.visualstudio.com/_packaging/ApexTestFramework/nuget/v3/index.json" -UserName $env:apexUsername -Password $env:apexPat
```
Comment (rohit21agrawal) [inline build/common.ps1]: change the source url into a variable too.

also use these variables:  VstsPackageFeedUrl , VstsPackageFeedUsername , VstsPackageFeedPassword

File: build/common.ps1 [hunk 4]
```diff
@@ -20,6 +20,8 @@ $DotNetExe = Join-Path $CLIRoot 'dotnet.exe'
 $NuGetExe = Join-Path $NuGetClientRoot '.nuget\nuget.exe'
 $XunitConsole = Join-Path $NuGetClientRoot 'packages\xunit.runner.console.2.1.0\tools\xunit.console.exe'
 $ILMerge = Join-Path $NuGetClientRoot 'packages\ILMerge.2.14.1208\tools\ILMerge.exe'
+$VstsPackageFeedUrl = 'https://devdiv.pkgs.visualstudio.com/_packaging/ApexTestFramework/nuget/v3/index.json'
```
Comment (rohit21agrawal) [inline build/common.ps1]: I meant assume this will be set as an environment variable, we don’t want to disclose private feeds so please force push

**Review by rohit21agrawal — CHANGES_REQUESTED** [review]

**Review by emgarten — APPROVED** [review]

File: build/build.proj [hunk 5]
```diff
@@ -122,6 +122,33 @@
...
   <!--
     ============================================================
+    Run VS15 apex tests
+    ============================================================
+  -->
+  <Target Name="ApexTestsStandalone">
+
+    <!-- Test inputs -->
+    <PropertyGroup>
+      <TestProjectPaths>$(RepositoryRootDirectory)test\NuGet.Tests.Apex\NuGet.Tests.Apex\NuGet.Tests.Apex.csproj</TestProjectPaths>
```
Comment (rohit21agrawal) [inline build/build.proj:132]: use @(ApexProjects) instead

File: build/common.ps1 [hunk 6]
```diff
@@ -189,6 +189,13 @@ Function Install-NuGet {
     & $NuGetExe locals all -list -verbosity detailed
 }
 
+Function Add-ApexFeed {
+    if (Test-Path $NuGetExe) {    
+        # Add Apex VSTS feed
+        & $NuGetExe sources add -Name ApexFeed -Source $env:VstsPackageFeedUrl -UserName $env:VstsPackageFeedUsername -Password $env:VstsPackageFeedPassword
```
Comment (rohit21agrawal) [inline build/common.ps1]: which config file will this get added to? the one in our repo or the one is %applocaldata% ?

File: configure.ps1 [hunk 7]
```diff
@@ -54,6 +54,12 @@ Invoke-BuildStep 'Installing .NET CLI' {
     Install-DotnetCLI -Force:$Force    
 } -ev +BuildErrors
 
+if ($env:Ci) {
```
Comment (rohit21agrawal) [inline configure.ps1]: nit: uppercase CI

File: build/build.proj [hunk 8]
```diff
@@ -337,6 +364,27 @@
 
   <!--
     ============================================================
+    Restore Apex 
+    ============================================================
+  -->
+  <Target Name="RestoreApex">
```
Comment (rohit21agrawal) [inline build/build.proj:370]: does the ApexTestsStandalone target need to depend on this target or both need to be run separately?

**Review by rohit21agrawal — APPROVED** [review]

**Review by rohit21agrawal — CHANGES_REQUESTED** [review]
remove the configure.ps1 changes as i am able to run apex tests on CI without it but please fix the SaveAs dialog that keeps on appearing.

I was able to run tests, here are the results:

https://devdiv.visualstudio.com/DevDiv/NuGet/_release?definitionId=401&releaseId=49463&_a=release-contribution-tab-ms.vss-test-web.test-result-in-release-management

**Comment by rohit21agrawal** [comment]
@zhili1208  can you check this in without the configure.ps1/common.ps1 changes?

**Pull request merged**